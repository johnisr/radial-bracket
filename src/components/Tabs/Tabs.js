import React from 'react';
import './Tabs.css';

export const TabList = (props) => {
  const { activeIndex } = props;
  const children = React.Children.map(props.children, (child, index) => {
    return React.cloneElement(child, {
      isActive: index === activeIndex,
      onActivate: () => {
        props.onActiveTab(index)
      }
    });
  });
  return (
    <div className="tabs">{children}</div>
  );
}

export const Tab = (props) => {
  const isDisabled = props.isDisabled;
  const isActive = props.isActive;
  const style = isDisabled ? "tab tab--disabled" :
    isActive ? "tab tab--active" : "tab";
  return (
    <div
      className={style}
      onClick={isDisabled ? null : () => props.onActivate()}
    >
      {props.children}
    </div>
  );
}

export const TabPanels = (props) => {
  const activeIndex = props.activeIndex;
  return (
    <div className="tabPanels">
      {props.children[activeIndex]}
    </div>
  );
}

export const TabPanel = (props) => (
  <div>
    {props.children}
  </div>
)

export class Tabs extends React.Component {
  constructor(props) {
    super(props);
    this.state = { 
      activeIndex: this.props.activeIndex ? this.props.activeIndex : 0
    };
  }
  selectTabIndex(activeIndex) {
    this.setState({ activeIndex });
  }
  componentWillReceiveProps = (nextProps) => {
    const { activeIndex } = nextProps;
    this.setState({ activeIndex });
  }
  
  render() {
    const activeIndex = this.state.activeIndex;
    const children = React.Children.map(this.props.children, (child) => {
      if (child.type === TabPanels) {
        return React.cloneElement(child, { activeIndex });
      } else if (child.type === TabList) {
        return React.cloneElement(child, {
          activeIndex,
          onActiveTab: (activeIndex) => {
            this.setState({ activeIndex });
          }
        });
      } else {
        return child;
      } 
    });
    return (
      <div>
        {children}
      </div>
    )
  }
}

export const DataTabs = (props) => {
  const { data } = props;
  return (
    <Tabs>
      <TabList>
        {data.map(tab => (
          <Tab>{tab.label}</Tab>
        ))}
      </TabList>
      <TabPanels>
        {data.map(tab => (
          <TabPanel>{tab.description}</TabPanel>
        ))}
      </TabPanels>
    </Tabs>
  );
};