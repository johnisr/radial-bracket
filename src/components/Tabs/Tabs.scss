@import '../../styles/base';

.tabs {
  display: flex;
  margin-top: 2rem;
  background-color: darken($color-grey-light-1, 10%);
}

.tab {
  display: inline-block;
  padding: 1rem 2rem;
  color: $color-grey-dark-2;
  background-color: darken($color-grey-light-1, 10%);
  cursor: pointer;
  font-family: $font-display;
  
  
  &:hover {
    background-color: $color-grey-light-2;
    
  }
  
  &:active {
    background-color: $color-grey-light-2;
  }
  
  &--active {
    background-color: $color-grey-light-2;
  }

  &--disabled {
    opacity: .4;
    cursor: auto;

    &:hover {
      background-color: $color-grey-light-1;
    }

    &:active {
      box-shadow: none;
    }
    
  }
}

.tabPanels {
  background-color: $color-grey-light-2;
  padding: 2rem;
}