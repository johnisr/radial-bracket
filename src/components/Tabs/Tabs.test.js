import React from 'react';
import { shallow } from 'enzyme';
import { TabList, Tab, TabPanels, TabPanel, Tabs } from './Tabs';

it('should render TabList', () => {
  const wrapper = shallow(<TabList
    activeIndex={0}
    children={
      <div>
        <Tab />
        <Tab />
      </div>
    }
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should render active Tab', () => {
  const wrapper = shallow(<Tab
    isDisabled={false}
    isActive={true}
    children={<p>Hello</p>}
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should render disabled Tab', () => {
  const wrapper = shallow(<Tab
    isDisabled={true}
    isActive={false}
    children={<p>Hello</p>}
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should render normal Tab', () => {
  const wrapper = shallow(<Tab
    isDisabled={false}
    isActive={false}
    children={<p>Hello</p>}
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should render TabPanels', () => {
  const wrapper = shallow(<TabPanels
    activeIndex={0}
    children={
      <div>
        <TabPanel />
        <TabPanel />
      </div>
    }
  />);
  expect(wrapper).toMatchSnapshot();
});

it('should render TabPanel', () => {
  const wrapper = shallow(<TabPanel children={<p>Hello</p>}/>);
  expect(wrapper).toMatchSnapshot();
});

it('should render Tabs', () => {
  const wrapper = shallow(<TabPanel activeIndex={0} children={
    <Tabs>
      <TabList>
        <Tab>Hello</Tab>
      </TabList>
      <TabPanels>
        <TabPanel>World</TabPanel>
      </TabPanels>
    </Tabs>
  }/>);
  expect(wrapper).toMatchSnapshot();
});