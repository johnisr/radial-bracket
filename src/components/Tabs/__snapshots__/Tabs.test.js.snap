// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TabList 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    activeIndex={0}
>
    <div>
        <Unknown />
        <Unknown />
    </div>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          isActive={true}
          onActivate={[Function]}
>
          <Unknown />
          <Unknown />
</div>,
      ],
      "className": "tabs",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": ".0",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Unknown />,
            <Unknown />,
          ],
          "isActive": true,
          "onActivate": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            isActive={true}
            onActivate={[Function]}
>
            <Unknown />
            <Unknown />
</div>,
        ],
        "className": "tabs",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": ".0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Unknown />,
              <Unknown />,
            ],
            "isActive": true,
            "onActivate": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render TabPanel 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown>
    <p>
        Hello
    </p>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Hello
</p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Hello",
      },
      "ref": null,
      "rendered": "Hello",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Hello
</p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hello",
        },
        "ref": null,
        "rendered": "Hello",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render TabPanels 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    activeIndex={0}
>
    <div>
        <Unknown />
        <Unknown />
    </div>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
      "className": "tabPanels",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "className": "tabPanels",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render Tabs 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    activeIndex={0}
>
    <Tabs>
        <Unknown>
            <Unknown>
                Hello
            </Unknown>
        </Unknown>
        <Unknown>
            <Unknown>
                World
            </Unknown>
        </Unknown>
    </Tabs>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Tabs>
        <Unknown>
                <Unknown>
                        Hello
                </Unknown>
        </Unknown>
        <Unknown>
                <Unknown>
                        World
                </Unknown>
        </Unknown>
</Tabs>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Unknown>
            <Unknown>
                        Hello
            </Unknown>
</Unknown>,
          <Unknown>
            <Unknown>
                        World
            </Unknown>
</Unknown>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Unknown>
              Hello
</Unknown>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Hello",
            },
            "ref": null,
            "rendered": "Hello",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Unknown>
              World
</Unknown>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "World",
            },
            "ref": null,
            "rendered": "World",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Tabs>
          <Unknown>
                    <Unknown>
                              Hello
                    </Unknown>
          </Unknown>
          <Unknown>
                    <Unknown>
                              World
                    </Unknown>
          </Unknown>
</Tabs>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Unknown>
              <Unknown>
                            Hello
              </Unknown>
</Unknown>,
            <Unknown>
              <Unknown>
                            World
              </Unknown>
</Unknown>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Unknown>
                Hello
</Unknown>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Hello",
              },
              "ref": null,
              "rendered": "Hello",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Unknown>
                World
</Unknown>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "World",
              },
              "ref": null,
              "rendered": "World",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render active Tab 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    isActive={true}
    isDisabled={false}
>
    <p>
        Hello
    </p>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Hello
</p>,
      "className": "tab tab--active",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Hello",
      },
      "ref": null,
      "rendered": "Hello",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Hello
</p>,
        "className": "tab tab--active",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hello",
        },
        "ref": null,
        "rendered": "Hello",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render disabled Tab 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    isActive={false}
    isDisabled={true}
>
    <p>
        Hello
    </p>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Hello
</p>,
      "className": "tab tab--disabled",
      "onClick": null,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Hello",
      },
      "ref": null,
      "rendered": "Hello",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Hello
</p>,
        "className": "tab tab--disabled",
        "onClick": null,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hello",
        },
        "ref": null,
        "rendered": "Hello",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`should render normal Tab 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    isActive={false}
    isDisabled={false}
>
    <p>
        Hello
    </p>
</Unknown>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Hello
</p>,
      "className": "tab",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Hello",
      },
      "ref": null,
      "rendered": "Hello",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Hello
</p>,
        "className": "tab",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hello",
        },
        "ref": null,
        "rendered": "Hello",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
